### PR TITLE
Mute bwc tests for old versions in FIPS JVMs

### LINKED ISSUE #32727: Create new secure setting type that is stored (encrypted) in cluster state
**Goal:**
There are some features ([one example](https://github.com/elastic/elasticsearch/pull/31087)) that could make use of a secure setting that is consistent across nodes.

**Problem with existing secure settings:**
Existing secure settings stored in the keystore have the problem that they are not enforced to be consistent across nodes. The key-store settings are primarily used for node start-up secure settings.

**Features that would enable such a setting**:
In order to have an arbitrary set of settings stored securely out in the open in the cluster-state, there must be a way to encrypt and decrypt these values. One solution would be to introduce a sort-of system key that will be stored in the secure keystore, and all these new cluster-state-secure-settings would use this key to encrypt and decrypt their values when needed. These settings probably won't change, and it may make sense to enforce that by disallowing updates to these settings.

**rotated system key**:
Although these secure settings may not need key-rotation, the master system key used to cypher these settings should have a way to rotate itself. This means having a hand-off mechanism where old key decrypts and new key encrypts and re-inserts into the cluster-state

I've left further implementation details out of this description, since those may be hashed out upon investigation

cc @elastic/es-distributed @elastic/es-security

### LINKED ISSUE #32737: :x-pack:qa:full-cluster-restart fails with SecurityException: Keystore has been corrupted or tampered with on FIPS JVM
Relevant log 
```
18:02:02 > Task :x-pack:qa:full-cluster-restart:with-system-key:v6.3.3-SNAPSHOT#oldClusterTestCluster#node0.addToKeystore#xpack.watcher.encryption_key
18:02:02 Task ':x-pack:qa:full-cluster-restart:with-system-key:v6.3.3-SNAPSHOT#oldClusterTestCluster#node0.addToKeystore#xpack.watcher.encryption_key' is not up-to-date because:
18:02:02   Task has not declared any outputs despite executing actions.
18:02:02 	at sun.security.provider.JavaKeyStore.engineLoad(JavaKeyStore.java:658)
18:02:02 	at sun.security.provider.JavaKeyStore$JKS.engineLoad(JavaKeyStore.java:56)
18:02:02 	at sun.security.provider.KeyStoreDelegator.engineLoad(KeyStoreDelegator.java:224)
18:02:02 	at sun.security.provider.JavaKeyStore$DualFormatJKS.engineLoad(JavaKeyStore.java:70)
18:02:02 	at java.security.KeyStore.load(KeyStore.java:1445)
18:02:02 Starting process 'command '/var/lib/jenkins/workspace/elastic+elasticsearch+master+matrix-java-periodic/ES_BUILD_JAVA/java11/ES_RUNTIME_JAVA/java8fips/nodes/virtual&&linux/x-pack/qa/full-cluster-restart/with-system-key/build/cluster/v6.3.3-SNAPSHOT#oldClusterTestCluster node0/elasticsearch-6.3.3-SNAPSHOT/bin/elasticsearch-keystore''. Working directory: /var/lib/jenkins/workspace/elastic+elasticsearch+master+matrix-java-periodic/ES_BUILD_JAVA/java11/ES_RUNTIME_JAVA/java8fips/nodes/virtual&&linux/x-pack/qa/full-cluster-restart/with-system-key/build/cluster/v6.3.3-SNAPSHOT#oldClusterTestCluster node0/cwd Command: /var/lib/jenkins/workspace/elastic+elasticsearch+master+matrix-java-periodic/ES_BUILD_JAVA/java11/ES_RUNTIME_JAVA/java8fips/nodes/virtual&&linux/x-pack/qa/full-cluster-restart/with-system-key/build/cluster/v6.3.3-SNAPSHOT#oldClusterTestCluster node0/elasticsearch-6.3.3-SNAPSHOT/bin/elasticsearch-keystore add-file xpack.watcher.encryption_key /var/lib/jenkins/workspace/elastic+elasticsearch+master+matrix-java-periodic/ES_BUILD_JAVA/java11/ES_RUNTIME_JAVA/java8fips/nodes/virtual&&linux/x-pack/qa/full-cluster-restart/src/test/resources/system_key
18:02:02 Successfully started process 'command '/var/lib/jenkins/workspace/elastic+elasticsearch+master+matrix-java-periodic/ES_BUILD_JAVA/java11/ES_RUNTIME_JAVA/java8fips/nodes/virtual&&linux/x-pack/qa/full-cluster-restart/with-system-key/build/cluster/v6.3.3-SNAPSHOT#oldClusterTestCluster node0/elasticsearch-6.3.3-SNAPSHOT/bin/elasticsearch-keystore''
18:02:02 warning: ignoring JAVA_TOOL_OPTIONS=-Dfile.encoding=UTF8
18:02:02 	at sun.security.util.AnchorCertificates$1.run(AnchorCertificates.java:61)
18:02:02 	at sun.security.util.AnchorCertificates$1.run(AnchorCertificates.java:52)
18:02:02 	at java.security.AccessController.doPrivileged(Native Method)
18:02:02 	at sun.security.util.AnchorCertificates.<clinit>(AnchorCertificates.java:52)
18:02:02 	at sun.security.provider.certpath.AlgorithmChecker.checkFingerprint(AlgorithmChecker.java:214)
18:02:02 	at sun.security.provider.certpath.AlgorithmChecker.<init>(AlgorithmChecker.java:164)
18:02:02 	at sun.security.provider.certpath.AlgorithmChecker.<init>(AlgorithmChecker.java:118)
18:02:02 	at sun.security.validator.SimpleValidator.engineValidate(SimpleValidator.java:157)
18:02:02 	at sun.security.validator.Validator.validate(Validator.java:262)
18:02:02 	at sun.security.validator.Validator.validate(Validator.java:238)
18:02:02 	at sun.security.validator.Validator.validate(Validator.java:207)
18:02:02 	at javax.crypto.JarVerifier.isTrusted(JarVerifier.java:610)
18:02:02 	at javax.crypto.JarVerifier.verifySingleJar(JarVerifier.java:530)
18:02:02 	at javax.crypto.JarVerifier.verifyJars(JarVerifier.java:363)
18:02:02 	at javax.crypto.JarVerifier.verify(JarVerifier.java:289)
18:02:02 	at javax.crypto.JceSecurity.verifyProviderJar(JceSecurity.java:164)
18:02:02 	at javax.crypto.JceSecurity.getVerificationResult(JceSecurity.java:190)
18:02:02 	at javax.crypto.JceSecurity.canUseProvider(JceSecurity.java:204)
18:02:02 	at javax.crypto.SecretKeyFactory.nextSpi(SecretKeyFactory.java:295)
18:02:02 	at javax.crypto.SecretKeyFactory.<init>(SecretKeyFactory.java:121)
18:02:02 	at javax.crypto.SecretKeyFactory.getInstance(SecretKeyFactory.java:160)
18:02:02 	at org.elasticsearch.common.settings.KeyStoreWrapper.createCipher(KeyStoreWrapper.java:288)
18:02:02 	at org.elasticsearch.common.settings.KeyStoreWrapper.decrypt(KeyStoreWrapper.java:336)
18:02:02 	at org.elasticsearch.common.settings.AddFileKeyStoreCommand.execute(AddFileKeyStoreCommand.java:73)
18:02:02 	at org.elasticsearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:86)
18:02:02 	at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:124)
18:02:02 	at org.elasticsearch.cli.MultiCommand.execute(MultiCommand.java:79)
18:02:02 	at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:124)
18:02:02 	at org.elasticsearch.cli.Command.main(Command.java:90)
18:02:02 	at org.elasticsearch.common.settings.KeyStoreCli.main(KeyStoreCli.java:41)
18:02:02 Exception in thread "main" java.lang.SecurityException: Keystore has been corrupted or tampered with
18:02:02 	at org.elasticsearch.common.settings.KeyStoreWrapper.decrypt(KeyStoreWrapper.java:349)
18:02:02 	at org.elasticsearch.common.settings.AddFileKeyStoreCommand.execute(AddFileKeyStoreCommand.java:73)
18:02:02 	at org.elasticsearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:86)
18:02:02 	at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:124)
18:02:02 	at org.elasticsearch.cli.MultiCommand.execute(MultiCommand.java:79)
18:02:02 	at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:124)
18:02:02 	at org.elasticsearch.cli.Command.main(Command.java:90)
18:02:02 	at org.elasticsearch.common.settings.KeyStoreCli.main(KeyStoreCli.java:41)
```

The  test is calling  `elasticsearch-keystore add xpack.watcher.encryption_key -x`

### PR DESCRIPTION
**Pull request by jkakavas**
Elasticsearch versions earlier than 6.4.0 cannot properly run in a
FIPS 140 JVM. This commit ensures that we do not try to run bwc
tests that entail spinning up < 6.4.0 nodes when CI is run in a FIPS
140 JVM.
It also reverts e497173 and e64bb48 as the workarounds inserted there
are no longer required.

Resolves #32737

### CODE REVIEW

File: x-pack/qa/full-cluster-restart/build.gradle [hunk 1]
```diff
@@ -129,8 +129,12 @@ subprojects {
       .file('src/test/resources/org/elasticsearch/xpack/security/transport/ssl/certs/simple/testnode.jks')
     into outputDir
   }
-
-  for (Version version : bwcVersions.indexCompatible) {
+  List<Version> indexCompatibleVersions = bwcVersions.indexCompatible
+  // Versions before 6.4.0 cannot be run in a FIPS 140 JVM, so exclude them
+  if (inFipsJvm){
+    indexCompatibleVersions.removeAll { it.before("6.4.0")}
```
Comment (atorok) [inline x-pack/qa/full-cluster-restart/build.gradle]: I would prefer this not to be repeated in the build logic and encapsulated in `VersionCollection`, perhaps by passing  `inFipsJvm` to  the constructor.

Comment (jkakavas) [reply x-pack/qa/full-cluster-restart/build.gradle]: When calling the constructor in `build.gradle`, the `inFipsJvm` property hasn't been evaluated, as this happens in `BuildPlugin#globalBuildInfo()`. 

Another approach would be to add a `inFipsJvm` optional param to `getSnapshotsIndexCompatible()` , `getIndexCompatible()` etc, defaulting to false. WDYT?

**Review by alpar-t — APPROVED** [review]

**Comment by colings86** [comment]
Just a quick note: I think the "Resolves" reference int eh description might be wrong as it links to an issue that seems unrelated?

**Review by rjernst — CHANGES_REQUESTED** [review]
I don't think this makes sense as something VersionCollection figures out. Whether a particular version will run on a fips JVM has to do with that version, not version compatibility, which is what VersionCollection is all about. Instead of doing this on each compatible list method, I think you can do it as an attribute on the Version object itself? Then the code in eg full cluster restart tests can be something like:

```
for (Version version : bwcVersions.wireCompatible) {
    if (inFipsJvm && version.isFipsCompatible() == false) {
        continue
    }
    ....
}
```

**Comment by jkakavas** [comment]
Thanks for the feedback @rjernst. I'm closing this PR in favor of #32901 ,that re-enables the tests and addresses the root cause instead.

**Pull request closed without merging**